src/backend/distributed/operations/shard_split.c

```diff
@@ -0,0 +1,1029 @@
+					   List *destinationWorkerNodesList)
+{
+	StringInfo splitCopyInfoArray = makeStringInfo();
+	appendStringInfo(splitCopyInfoArray, "ARRAY[");
+
+	ShardInterval *splitChildShardInterval = NULL;
+	bool addComma = false;
+	WorkerNode *destinationWorkerNode = NULL;
+	forboth_ptr(splitChildShardInterval, splitChildrenShardIntervalList,
+				destinationWorkerNode, destinationWorkerNodesList)
+	{
+		if (addComma)
+		{
+			appendStringInfo(splitCopyInfoArray, ",");
+		}
+
+		StringInfo splitCopyInfoRow = makeStringInfo();
+		appendStringInfo(splitCopyInfoRow,
+						 "ROW(%lu, %d, %d, %u)::citus.split_copy_info",
+						 splitChildShardInterval->shardId,
+						 DatumGetInt32(splitChildShardInterval->minValue),
+						 DatumGetInt32(splitChildShardInterval->maxValue),
+						 destinationWorkerNode->nodeId);
+		appendStringInfo(splitCopyInfoArray, "%s", splitCopyInfoRow->data);
+
+		addComma = true;
+	}
+	appendStringInfo(splitCopyInfoArray, "]");
+
+	StringInfo splitCopyUdf = makeStringInfo();
+	appendStringInfo(splitCopyUdf, "SELECT pg_catalog.worker_split_copy(%lu, %s);",
+					 sourceShardSplitInterval->shardId,
+					 splitCopyInfoArray->data);
+
+	return splitCopyUdf;
+}
+
+
+/*
+ * Create an object on a worker node.
+ */
+static void
+CreateObjectOnPlacement(List *objectCreationCommandList,
+						WorkerNode *workerPlacementNode)
+{
+	char *currentUser = CurrentUserName();
+	SendCommandListToWorkerOutsideTransaction(workerPlacementNode->workerName,
+											  workerPlacementNode->workerPort,
+											  currentUser,
+											  objectCreationCommandList);
+}
+
+
+/*
+ * Create split children intervals for a shardgroup given list of split points.
+ * Example:
+ * 'sourceColocatedShardIntervalList': Colocated shard S1[-2147483648, 2147483647] & S2[-2147483648, 2147483647]
+ * 'splitPointsForShard': [0] (2 way split)
+ * 'shardGroupSplitIntervalListList':
+ *  [
+ *      [ S1_1(-2147483648, 0), S1_2(1, 2147483647) ], // Split Interval List for S1.
+ *      [ S2_1(-2147483648, 0), S2_2(1, 2147483647) ]  // Split Interval List for S2.
+ *  ]
+ */
+static List *
+CreateSplitIntervalsForShardGroup(List *sourceColocatedShardIntervalList,
+								  List *splitPointsForShard)
+{
+	List *shardGroupSplitIntervalListList = NIL;
+
+	ShardInterval *shardToSplitInterval = NULL;
+	foreach_ptr(shardToSplitInterval, sourceColocatedShardIntervalList)
+	{
+		List *shardSplitIntervalList = NIL;
+		CreateSplitIntervalsForShard(shardToSplitInterval, splitPointsForShard,
+									 &shardSplitIntervalList);
+
+		shardGroupSplitIntervalListList = lappend(shardGroupSplitIntervalListList,
+												  shardSplitIntervalList);
+	}
+
+	return shardGroupSplitIntervalListList;
+}
+
+
+/*
+ * Create split children intervals given a sourceshard and a list of split points.
+ * Example: SourceShard is range [0, 100] and SplitPoints are (15, 30) will give us:
+ *  [(0, 15) (16, 30) (31, 100)]
+ */
+static void
+CreateSplitIntervalsForShard(ShardInterval *sourceShard,
```

Comment:
Given that we can practically create unlimited number of shards with this API, should we add an upper limit check on the shard count: `MAX_SHARD_COUNT` (shared_library_init.h) per table? 

I don't think there are any hard limits on the number of shards per table, but this is more soft limit where above that limit is not very well tested in terms of metadata cache management (e..g,  per backend metadata cache could become giant causing lots of OOMs / stability issues)


Query to generate lots of shards:
```SQL

 WITH shard_ranges AS (SELECT ((1879048192 + i))::text as r, 7 as n FROM generate_series(0,134217727) i )

SELECT citus_split_shard_by_split_points(102038, array_agg(r), array_agg(n) || 7) FROM shard_ranges;
 
```

Reply:
After creating `~13400` shards, multi-shard queries becomes pretty slow, so probably good to have some upper limit:
```SQL
select count(*) from test ;
┌───────┐
│ count │
├───────┤
│     1 │
└───────┘
(1 row)

Time: 41961.383 ms (00:41.961)
```